src/Features/Core/Portable/EditAndContinue/EditSession.cs

```diff
@@ -4,116 +4,145 @@
-        private EncEditSessionInfo _encEditSessionInfo = new EncEditSessionInfo();
+        /// <summary>
+        /// A <see cref="DocumentId"/> is added whenever <see cref="EditAndContinueDiagnosticAnalyzer"/> reports 
+        /// rude edits or module diagnostics. At the end of the session we ask the diagnostic analyzer to reanalyze 
+        /// the documents to clean up the diagnostics.
+        /// </summary>
+        private readonly HashSet<DocumentId> _documentsWithReportedDiagnostics;
+        private readonly object _documentsWithReportedDiagnosticsGuard = new object();
 
-        private readonly ImmutableDictionary<ActiveMethodId, ImmutableArray<NonRemappableRegion>> _nonRemappableRegions;
+        private bool _changesApplied;
 
-        internal EditSession(
-            Solution baseSolution,
-            DebuggingSession debuggingSession,
-            IActiveStatementProvider activeStatementProvider,
-            ImmutableDictionary<ProjectId, ProjectReadOnlyReason> projects,
-            ImmutableDictionary<ActiveMethodId, ImmutableArray<NonRemappableRegion>> nonRemappableRegions,
-            bool stoppedAtException)
+        internal EditSession(DebuggingSession debuggingSession, EditSessionTelemetry telemetry, bool stoppedAtException)
         {
-            Debug.Assert(baseSolution != null);
             Debug.Assert(debuggingSession != null);
-            Debug.Assert(activeStatementProvider != null);
-            Debug.Assert(nonRemappableRegions != null);
+            Debug.Assert(telemetry != null);
 
-            _baseSolution = baseSolution;
-            _debuggingSession = debuggingSession;
-            _activeStatementProvider = activeStatementProvider;
-            _stoppedAtException = stoppedAtException;
-            _projects = projects;
-            _cancellation = new CancellationTokenSource();
+            DebuggingSession = debuggingSession;
 
-            // TODO: small dict, pool?
-            _analyses = new Dictionary<DocumentId, Analysis>();
+            _analyses = new Dictionary<DocumentId, (Document Document, AsyncLazy<DocumentAnalysisResults> Results)>();
+            _documentsWithReportedDiagnostics = new HashSet<DocumentId>();
+            _moduleDiagnostics = new Dictionary<Guid, ImmutableArray<LocationlessDiagnostic>>();
 
-            // TODO: small dict, pool?
-            _documentsWithReportedRudeEdits = new HashSet<DocumentId>();
+            _nonRemappableRegions = debuggingSession.NonRemappableRegions;
 
-            _nonRemappableRegions = nonRemappableRegions;
+            Telemetry = telemetry;
+            BaseSolution = debuggingSession.LastCommittedSolution;
 
             BaseActiveStatements = new AsyncLazy<ActiveStatementsMap>(GetBaseActiveStatementsAsync, cacheResult: true);
             BaseActiveExceptionRegions = new AsyncLazy<ImmutableArray<ActiveStatementExceptionRegions>>(GetBaseActiveExceptionRegionsAsync, cacheResult: true);
```

Comment:
❔ It seems "obvious" to me that if `cacheResult` is omitted, it will be cached. Is this not the case for the type? If it is cached by default, this parameter could be removed without hurting readability.

Reply:
The parameter is not optional.

---
In reply to: [270626486](https://github.com/dotnet/roslyn/pull/34027#discussion_r270626486) [](ancestors = 270626486)